### PR TITLE
ARROW-9937: [Rust] [DataFusion] Improved aggregations

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR is a proposal to fix 4 issues in our aggregations:

1. averages are incorrect
2. they only support aggregations that can be reduced using a single value (reason for issue 1.)
3. they do not leverage arrow’s aggregate kernels nor memory layout
4. they only support a single column

The proposal is written here: https://docs.google.com/document/d/1n-GS103ih3QIeQMbf_zyDStjUmryRQd45ypgk884LHU/edit?usp=sharing

Its main features:

1. adds a test of a wrong average and fixes it
1. makes `ScalarValue` a nullable dynamic type, which is closer to how `Array` works (it is nullable)
2. Accumulators now know how to be updated from values (partial) and from other accumulators' state (final)
3. Accumulators can now receive more than one column
4. AggregateExec now knows how to serialize aggregators' state into a ArrayRef's, so that they can be passed throughout the execution
5. Aggregations are now always made in two steps: partial (update from values) and full (update from other's states)
6. Aggregations leverage arrow's kernels as much as possible (all aggregates + take + concatenate)

This PR is built on top of 3 PRs that are under review, and thus is only a draft at this point. 

The benchmarks are between -30% and +15% on my computer (AFAIK no SIMD). Given that the computation now always requires two passes, I was sufficiently happy with them. More can be achieved later.

I am still evaluating the reason for the `aggregate_query_group_by`, but given the functionality that it adds, I considered it sufficiently good for some initial discussions, @andygrove , @nevi-me , @alamb @paddyhoran .

The benchmarks were updated to better reflect real data, and the results are as follows:

```
aggregate_query_no_group_by 15 12                                                                            
                        time:   [478.23 us 479.62 us 480.98 us]
                        change: [-29.686% -27.511% -25.784%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe

aggregate_query_group_by 15 12                                                                             
                        time:   [2.8689 ms 2.8794 ms 2.8922 ms]
                        change: [+12.971% +13.710% +14.445%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

aggregate_query_group_by_with_filter 15 12                                                                             
                        time:   [2.1803 ms 2.2062 ms 2.2330 ms]
                        change: [-8.2400% -6.7872% -5.3209%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
```

Sorry for the long PR, but this was a relatively difficult PR to achieve, as it required refactoring of some of our most delicate components. I will try to split it in smaller parts to each the review.

### CODE REVIEW

**Comment by alamb** [comment]
I plan to review this carefully tomorrow morning US eastern time

File: rust/datafusion/benches/aggregate_query_sql.rs
Comment (jorgecarleitao) [inline rust/datafusion/benches/aggregate_query_sql.rs:85]: This should be random, i%4 is quite predictable.

File: rust/datafusion/benches/aggregate_query_sql.rs
Comment (jorgecarleitao) [inline rust/datafusion/benches/aggregate_query_sql.rs:121]: 11, not 12.

File: rust/datafusion/src/physical_plan/common.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/common.rs:297]: So far this was enough because we do not have aggregations with other types, but this is needed because some batches can have no entries, in which case we need to build an empty record batch.

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/expressions.rs:107]: `AggregateExpr` has this information with them now because it allows them to `create_accumulator` without access to the input schema. This is helpful because, on the second pass, we need to create accumulators on the fly, and the input_schema of the second pass is different, as it now corresponds to the schema 

`[group1, group2, agg1_state1, agg1_state2, ...]`

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/expressions.rs:146]: This was operating on a row-by-row basis, which was replaced by a batch operation using `compute::sum`.

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/expressions.rs:402]: This is the prime example of this PR: the merge here uses two states to change two states from the accumulator.

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/expressions.rs:1839]: I removed these because the types are now selected from the signature, and no longer inferred by `AggregateExpr`.

File: rust/datafusion/src/execution/context.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/execution/context.rs:852]: A test for grouped with an empty result

File: rust/datafusion/src/execution/context.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/execution/context.rs:1199]: This test fails in master, with 3.25 != 3.0

File: rust/datafusion/src/physical_plan/hash_aggregate.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:289]: wrong comment, 1.2 and 1.3 are together.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:669]: I placed controlled data here so that we cab easily check the result

File: rust/datafusion/src/physical_plan/planner.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/planner.rs:236]: This seems major for a single partition.

File: rust/datafusion/src/scalar.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/scalar.rs:30]: no longer needed?

File: rust/arrow/src/compute/kernels/aggregate.rs
Comment (alamb) [inline rust/arrow/src/compute/kernels/aggregate.rs:37]: so the idea here is that the code is faster if we don't have to check each element for nulls?

File: rust/datafusion/src/physical_plan/aggregates.rs
Comment (alamb) [inline rust/datafusion/src/physical_plan/aggregates.rs:149]: This is fancy -- do I correctly understand that it allows multiple columns in an aggregate expression (like `MAX(c1, c2)...`)?

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (alamb) [inline rust/datafusion/src/physical_plan/expressions.rs:191]: 👍

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (alamb) [reply rust/datafusion/src/physical_plan/expressions.rs:1839]: makes sense

File: rust/datafusion/src/physical_plan/planner.rs
Comment (alamb) [reply rust/datafusion/src/physical_plan/planner.rs:236]: I a not sure I understand what you are getting at with this comment

**Review by alamb — COMMENTED** [review]
I didn't fully grok / review all the code in this PR, but at a high level I like the structure. Nice work @jorgecarleitao 

The idea of two phase aggregates (aka compute average not by averaging averages, but by computing sums of sums and counts of counts and then dividing at the end) 👍 . 

The idea of using arrow compute kernels also seems good to me

I think I saw that this PR adds support for multiple columns in an aggregate expression (like `MAX(c1, c2)...`)? Is that true? 


Thoughts on how to split this up for easier review:

Perhaps https://github.com/apache/arrow/pull/8172/commits/93913d71e0736d41057182f686e8c4ea6210557c (benchmark) could be pulled into its own PR

You maybe could also split out the move of the code for scalar functions to `scalar.rs` in its own PR (with no functional changes) which would be an easy review and reduce the size of this PR substantially.

Then a follow on PR to add support for Nullable scalar value constants

I hear you regarding the PR backup and juggling several in-flight PRs, which is painful. Which ones are you waiting on? Do they need review? Perhaps I can re-review them or @andygrove  can provide some better visibility into what the path is for getting them approved / rejected.

**Comment by alamb** [comment]
BTW https://docs.google.com/document/d/1n-GS103ih3QIeQMbf_zyDStjUmryRQd45ypgk884LHU/edit# is a good read too -- thank you for writing that down

File: rust/datafusion/src/physical_plan/planner.rs
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/planner.rs:236]: Sorry. What I meant is that compared to the previous implementation, we double the amount of work, even when we did not need because there is a single partition.

IMO we should handle this in another way, e.g. via a physical optimizer that re-writes the partial aggregation by a final aggregation when the input is a single partition.

File: rust/arrow/src/compute/kernels/aggregate.rs
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/aggregate.rs:37]: yes. Generally, operations on non-null fields are faster because there isn't an unpredictable branch on the loop. This check removes that unpredictable branch altogether when there are no nulls.

We could probably still get some more juice by using some of the vertical operations supported by packed_simd.

File: rust/datafusion/src/physical_plan/aggregates.rs
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/aggregates.rs:149]: I have not though about that, but that is an interesting idea 👍 

In this PR, `max` continues to only support a single column, which we select in [this line](https://github.com/apache/arrow/pull/8172/files#diff-a98d5d588d3c5b525c6840271a5bdddcR571).

This PR does enable us to create aggregate functions with more than one argument, and therefore this allows that option if we wish so. My initial thinking was supporting aggregate functions of more arguments just to support things like `covariance` and `correlation`, but now that you mention, we can do a lot of other things also. Another one is count distinct over N columns.

File: rust/datafusion/src/physical_plan/planner.rs
Comment (andygrove) [reply rust/datafusion/src/physical_plan/planner.rs:236]: I agree, the planner/optimizer should prevent this.

**Comment by andygrove** [comment]
I did a quick skim through of this and LGTM so far. I will have more time available during the coming week to review more thoroughly and pull the PR locally for some testing.

**Comment by jorgecarleitao** [comment]
@alamb , I tried to take the ScalarValue out of the PR for over 2hs, but I was unable: there is too much dependency on it on the current accumulators, and it would be a lot of new code that would then be again removed by this PR.

**Comment by alamb** [comment]
Thanks @jorgecarleitao  -- I don't feel strongly about the ScalarValue thign

File: rust/datafusion/src/sql/planner.rs
Comment (andygrove) [inline rust/datafusion/src/sql/planner.rs:346]: Could we math on `Some(n)` or add `if n.is_some()` on the match arm to be sure we aren't calling `unwrap` on ` None`?

**Comment by andygrove** [comment]
I've been testing with the TPC-H benchmark this morning.

Latest from master:

```
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 24, batch_size: 4096, path: "/mnt/tpch/parquet/100-240", file_format: "parquet" }
Query 1 iteration 0 took 16167 ms
Query 1 iteration 1 took 16012 ms
Query 1 iteration 2 took 16229 ms
```

This PR:

```
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 24, batch_size: 4096, path: "/mnt/tpch/parquet/100-240", file_format: "parquet" }
Query 1 iteration 0 took 13638 ms
Query 1 iteration 1 took 13559 ms
Query 1 iteration 2 took 13622 ms
```

That's around 15% speedup :rocket:

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @jorgecarleitao

**Comment by andygrove** [comment]
@jorgecarleitao This needs rebasing against master. I ran into compilation issues after merging master into this branch.

```
error[E0308]: mismatched types
   --> datafusion/src/physical_plan/functions.rs:504:33
    |
504 |             ScalarValue::UInt64(1),
    |                                 ^
    |                                 |
    |                                 expected enum `std::option::Option`, found integer
    |                                 help: try using a variant of the expected enum: `Some(1)`
```

**Comment by jorgecarleitao** [comment]
@andygrove , that is great news! Really good to know that this stands a stronger benchmark. Thanks a lot for taking the time to run it.

I rebased against master and fixed the unwrap. Let's see if the CI passes.

**Pull request closed without merging**
